airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/components.py

```diff
@@ -0,0 +1,313 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+
+import datetime
+from collections import defaultdict
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import Any, Callable, Iterable, List, Mapping, MutableMapping, Optional, Union
+from urllib.parse import urlencode, urljoin
+
+import pendulum
+import requests
+from airbyte_cdk.sources.declarative.extractors.record_extractor import RecordExtractor
+from airbyte_cdk.sources.declarative.extractors.record_filter import RecordFilter
+from airbyte_cdk.sources.declarative.incremental import CursorFactory, DatetimeBasedCursor, PerPartitionCursor
+from airbyte_cdk.sources.declarative.interpolation import InterpolatedString
+from airbyte_cdk.sources.declarative.partition_routers import CartesianProductStreamSlicer
+from airbyte_cdk.sources.declarative.partition_routers.single_partition_router import SinglePartitionRouter
+from airbyte_cdk.sources.declarative.requesters import HttpRequester
+from airbyte_cdk.sources.declarative.requesters.request_options.interpolated_request_options_provider import (
+    InterpolatedRequestOptionsProvider,
+    RequestInput,
+)
+from airbyte_cdk.sources.declarative.retrievers import SimpleRetriever
+from airbyte_cdk.sources.declarative.stream_slicers.stream_slicer import StreamSlicer
+from airbyte_cdk.sources.declarative.transformations import AddFields
+from airbyte_cdk.sources.declarative.transformations.add_fields import AddedFieldDefinition
+from airbyte_cdk.sources.declarative.types import Config, Record, StreamSlice, StreamState
+from airbyte_cdk.sources.streams.core import StreamData
+from airbyte_cdk.sources.streams.http import HttpClient
+from airbyte_cdk.sources.streams.http.exceptions import DefaultBackoffException, RequestBodyException, UserDefinedBackoffException
+from airbyte_cdk.sources.streams.http.http import BODY_REQUEST_METHODS
+from isodate import Duration, parse_duration
+
+from .utils import ANALYTICS_FIELDS_V2, FIELDS_CHUNK_SIZE, transform_data
+
+
+class SafeHttpClient(HttpClient):
```

Comment:
@lazebnyi can you describe why each of these components need to be custom and what their behavior is exactly? It should be clear from a reader next year why the connector cannot use the built-in implementations

Reply:
Added details in docs